### PR TITLE
`Programming exercises`: Allow tutors to assess submissions in the new code editor

### PR DESCRIPTION
**Pull request by pzdr7**
#### General
- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/#naming-conventions-for-github-pull-requests).

#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data.
- [x] I **strictly** followed the [client coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] Following the [theming guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.

#### Changes affecting Programming Exercises
- [x] **High priority**: I tested **all** changes and their related features with **all** corresponding user types on a test server configured with the **integrated lifecycle setup** (LocalVC and LocalCI).

### Motivation and Context

This should be the final view we have to replace until we can remove Ace from the online code editor.

### Description
- Introduces two new types of editor elements:
  - Line highlights
  - Glyph margin hover button (follows the mouse)
- Makes it possible to manually assess submissions using the new code editor

**Known issue**:

- around 70 lines of gradlew.bat in the java template are falsely highlighted as new. I suspect this has something to do with line endings, but the problem occurs with Ace as well.
 
### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Student
- 1 Programming Exercise
  - with a submission by the student and
  - with the online code editor enabled and
  - with manual assessment on (set the due date accordingly)

**As the student**

1. If you haven't done so already, make a submission to the programming exercise.

**As the instructor**

2. Sign in
3. Go to course management > your course > assessment > your programming exercise > exercise dashboard
4. Assess a new submission. Ensure that you can:

- Add a new line feedback item 
- Remove an existing line feedback item
- Edit an existing line feedback item
- Save your assessment.
- Override your assessment.

5. Make sure to set the assessment due date so the students can see the results

**As the student**

6. Sign in. 
7. View the assessment in the online code editor.
8. Ensure you can see all of the feedback you left.

#### Performance Review
- [x] I (as a reviewer) confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance 

#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

#### Client

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| [code-editor-monaco.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6419/latest/artifact/shared/Coverage-Report-Client-Tests/app/exercises/programming/shared/code-editor/monaco/code-editor-monaco.component.ts.html) | 91.44% | ✅ |
| [monaco-editor-glyph-margin-hover-button.model.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6419/latest/artifact/shared/Coverage-Report-Client-Tests/app/shared/monaco-editor/model/monaco-editor-glyph-margin-hover-button.model.ts.html) | 90.62% | ✅ |
| [monaco-editor-line-highlight.model.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6419/latest/artifact/shared/Coverage-Report-Client-Tests/app/shared/monaco-editor/model/monaco-editor-line-highlight.model.ts.html) | 100% | ✅ |
| [monaco-editor.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6419/latest/artifact/shared/Coverage-Report-Client-Tests/app/shared/monaco-editor/monaco-editor.component.ts.html) | 97.84% | ✅ |

### Screenshots

image

## Summary by CodeRabbit

- **New Features**
	- Enhanced code editor with new feedback suggestion system, allowing users to accept or discard suggestions.
	- Added highlighting of changed lines in the code editor for better visibility of modifications.
	- Implemented new user interface elements for feedback in the Monaco editor, including hover buttons and customizable line highlighting.
- **Improvements**
	- Updated code editor components to use Monaco editor consistently, improving performance and user interaction.
	- Added new styling options for Monaco editor to support both dark and default themes.
- **Bug Fixes**
	- Fixed issues related to file loading and event handling in the code editor.
- **Documentation**
	- Updated documentation to reflect new features and modifications in the code editor assessment process.
- **Tests**
	- Enhanced testing suite with new scenarios for feedback handling and editor interactions.
	- Adjusted existing tests to accommodate changes in editor components and assessment pages.

### CODE REVIEW

**Comment by pzdr7** [comment]
- Online Code Editor button doesn't work in score
- See if you can stop unsaved feedback from vanishing when save is clicked on the assessment

**Review by egekurt123 — DISMISSED** [review]
Tested manually on ts1 during testing session, works as expected

**Review by BaumiCoder — DISMISSED** [review]
Works fine in the testing session.

**Review by coolchock — APPROVED** [review]
tested during testing session, works as expected

**Review by b-fein — COMMENTED** [review]
Found an edge case that does not work:

- Submit a file that only has a single line.
- Two problems:
  - the :heavy_plus_sign: icon in the gutter is only shown when entering, leaving, and entering the editor area again with the mouse
  - adding a feedback does not show it

**Comment by pzdr7** [comment]
> Found an edge case that does not work:
> 
> * Submit a file that only has a single line.
> * Two problems:
>   
>   * the ➕ icon in the gutter is only shown when entering, leaving, and entering the editor area again with the mouse
>   * adding a feedback does not show it

Good catch! I'll adjust it.

**Review by b-fein — COMMENTED** [review]
Additional issue found: I cannot create an inline comment for a line that has a static code analysis issue. The :heavy_plus_sign: is not available since the :warning: is blocking it.

---

Possible enhancement: The contrast between the green background and the warning icon is pretty low in light mode:
image

**Comment by pzdr7** [comment]
@b-fein I've switched the add feedback button over to the left lane of the glyph margin. It is rendered to the left of any build annotations (SCA/build error) if necessary. What do you think?

**Review by b-fein — COMMENTED** [review]
Thanks for the quick fixes! Adding a feedback in first lines now works. Adding a comment in lines with static code analysis feedback also looks good now.

---

One additional issue:
The old editor was changed at some point to immediately put the keyboard focus on the text area field of the feedback (#6797). This behaviour is very convenient. It no longer works in the new editor.

**Comment by pzdr7** [comment]
I figured it would work because I saw it working on edit 😅. Should work for new feedback nodes now

**Review by b-fein — DISMISSED** [review]
Previously found issues have all been resolved. Thanks!

Code looks good, too.

**Review by b-fein — APPROVED** [review]

**Comment by coolchock** [comment]
re-tested on ts1, mentioned issues were resolved

**Review by coolchock — APPROVED** [review]

File: src/main/webapp/app/exercises/programming/assess/code-editor-tutor-assessment-container.component.ts [hunk 1]
```diff
@@ -295,20 +295,18 @@ export class CodeEditorTutorAssessmentContainerComponent implements OnInit, OnDe
...
      */
-    onFileLoad(selectedFile: string): void {
-        if (selectedFile && this.codeEditorContainer?.selectedFile && this.codeEditorContainer.aceEditor) {
-            // When the selectedFile is not part of the template, then this is a new file and all lines in code editor are highlighted
+    highlightChangedLines(selectedFile: string) {
+        if (selectedFile && this.codeEditorContainer?.selectedFile) {
             if (!this.templateFileSession[selectedFile]) {
-                const lastLine = this.codeEditorContainer.aceEditor.editorSession.getLength() - 1;
+                const lastLine = this.codeEditorContainer.getNumberOfLines();
                 this.highlightLines(0, lastLine);
```
[the remark this answers is not in the document]
Comment (pzdr7) [reply src/main/webapp/app/exercises/programming/assess/code-editor-tutor-assessment-container.component.ts:306]: In this file, the lines are supposed to be 0-based. I'll reduce the `lastLine` variable by 1 👍

**Review by coolchock — APPROVED** [review]
re-approve

**Review by BaumiCoder — APPROVED** [review]
Re-tested on my local machine

**Pull request merged**